runtime/compiler/trj9/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -15809,12 +15809,34 @@ J9::Z::TreeEvaluator::BCDCHKEvaluator(TR::Node * node, TR::CodeGenerator * cg)
                break;
                }
 
-            default: TR_ASSERT_FATAL(0, "BCDCHKEvaluator: Could not find recognized method for variable precision DAA operation.\n");
+            default:
+               {
+               /**
+                *
+                * BCDCHK can have a call node if the PD operation can be simplified to a No-Op. For example, one can get an integer via a call
+                * perform a i2pd followed by a pd2i. The pd2i (under BCDCHK) can be simplified to the icall.
+                * If this is the case, the lcall/icall must have been evaluated. We can skip the BCDCHK evaluation and return the call result.
+                */
+               TR_ASSERT_FATAL(resultReg != NULL && resultReg->getKind() == TR_GPR,
+                               "BCDCHKEvaluator: variable precision path encounters an unrecognized and unevaluated long/int call\n");
+               }
             }
          break;
          }
 
-      default: TR_ASSERT_FATAL(0, "Operation Code not supported by BCDCHKEvaluator.\n");
+      default:
+         {
+         /**
+          * Unrecognized opCodes under BCDCHK should come from optimizations such as local CSE and tree simplifications.
+          * They should be commoned nodes that's evaluated previously. Skip these nodes.
+          */
+         TR_ASSERT_FATAL(resultReg != NULL, "BCDCHKEvaluator: BCDCHK has an unevaluated non-PD node %p (non-PD op code %s) \n",
```

Comment:
Shouldn't it be the simplifier's job to remove the `BCDCHK` as well if it folds the operation underneath? This seems like special casing a situation that should never occur. The `BCDCHK` IL has a tight specification that it's first child is an operation that may raise a hardware exception. If a `BCDCHK` reaches codegen and this is not the case then the tree verifier (currently does not run in production builds) should have asserted that this is invalid IL.

Reply:
In theory, yes we can make the simplifier remove such `BCDCHK` nodes. There's one catch: the call parameter nodes under the `BCDCHK` has to be anchored somewhere and get evaluated. This is especially important for the single-reference-unevaluated call parameter nodes.

Shall we open an issue for the simplifier?

Reply:
> Shall we open an issue for the simplifier?

@NigelYiboYu yes I think so. Please reference this assert in the item you open. This assertion should happen at the tree verification level, i.e. check first child of `BCDCHK` and make sure it is a BCD operation.